telemetry/telemetry.go

```diff
@@ -29,48 +31,80 @@ func NewTelemetry(cfg *Config) *Telemetry {
 	}
 	t := &Telemetry{
 		Path:    "/metrics", // TODO hard-coded?
-		lock:    sync.RWMutex{},
 		Sensors: []*Sensor{},
 	}
 	t.addr = cfg.addr
-	t.mux = http.NewServeMux()
-	t.mux.Handle(t.Path, prometheus.Handler())
+	router := http.NewServeMux()
+	router.Handle(t.Path, prometheus.Handler())
+	t.Handler = router
+
 	for _, sensorCfg := range cfg.SensorConfigs {
 		sensor := NewSensor(sensorCfg)
 		t.Sensors = append(t.Sensors, sensor)
 	}
+	t.Rx = make(chan events.Event, 10)
 	return t
 }
 
-var listener net.Listener
+// Run executes the event loop for the telemetry server
+func (t *Telemetry) Run(bus *events.EventBus) {
+	t.Subscribe(bus, true)
+	t.Bus = bus
+	t.Start()
 
-// Serve starts serving the telemetry service
-func (t *Telemetry) Serve() {
-	t.lock.Lock()
-	defer t.lock.Unlock()
+	go func() {
+		defer t.Stop()
+		for {
+			event := <-t.Rx
+			switch event {
+			case
+				events.QuitByClose,
+				events.GlobalShutdown:
+				return
+			}
+		}
+	}()
+}
 
-	// No-op if we've created the server previously.
-	// TODO: golang's native implementation of http.Server.Server() cannot
-	// support graceful reload. We need to select an alternate implementation
-	// but in the meantime we need to back-out the change to reloading
-	// ref https://github.com/joyent/containerpilot/pull/165
-	if listener != nil {
-		return
-	}
-	ln, err := net.Listen(t.addr.Network(), t.addr.String())
-	if err != nil {
-		log.Fatalf("Error serving telemetry on %s: %v", t.addr.String(), err)
-	}
-	listener = ln
-	t.listening = true
+// Start starts serving the telemetry service
+func (t *Telemetry) Start() {
+	ln := t.listenWithRetry()
 	go func() {
-		log.Infof("telemetry: Listening on %s", t.addr.String())
-		log.Fatal(http.Serve(listener, t.mux))
-		log.Debugf("telemetry: Stopped listening on %s", t.addr.String())
+		log.Infof("telemetry: serving at %s", t.Addr)
+		t.Serve(ln)
+		log.Debugf("telemetry: stopped serving at %s", t.Addr)
 	}()
 }
 
-// Shutdown shuts down the telemetry service
-func (t *Telemetry) Shutdown() {
-	log.Debug("telemetry: shutdown received but currently a no-op")
+// on a reload we can't guarantee that the control server will be shut down
+// and the socket file cleaned up before we're ready to start again, so we'll
+// retry with the listener a few times before bailing out.
+func (t *Telemetry) listenWithRetry() net.Listener {
+	var (
+		err error
+		ln  net.Listener
+	)
+	for i := 0; i < 10; i++ {
+		ln, err = net.Listen(t.addr.Network(), t.addr.String())
+		if err == nil {
+			return ln
+		}
+		time.Sleep(time.Second)
+	}
+	log.Fatalf("error listening to socket at %s: %v", t.Addr, err)
+	return nil
+}
+
+// Stop shuts down the telemetry service
+func (t *Telemetry) Stop() {
+	log.Debug("telemetry: stopping server")
+	ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
```

Comment:
We _probably_ don't need it on the Telemetry server (although it can't really hurt either), but I think we want to avoid getting an error message back on the Control server. Otherwise the `/bin/containerpilot -reload` will show errors frequently when not warranted.